src/sage/graphs/matching_covered_graph.py

```diff
@@ -0,0 +1,1228 @@
+
+            sage: P = graphs.PetersenGraph()
+            sage: G = MatchingCoveredGraph(P)
+            sage: G.add_edge(1, 4)
+            sage: G.edges(sort=False)
+            [(0, 1, None), (0, 4, None), (0, 5, None), (1, 2, None), (1, 4, None), (1, 6, None), (2, 3, None), (2, 7, None), (3, 4, None), (3, 8, None), (4, 9, None), (5, 7, None), (5, 8, None), (6, 8, None), (6, 9, None), (7, 9, None)]
+
+        Adding an edge with both the incident vertices being existent such
+        that the resulting graph is not matching covered::
+
+            sage: C = graphs.CycleGraph(4)
+            sage: G = MatchingCoveredGraph(C)
+            sage: G.add_edge(0, 2)
+            Traceback (most recent call last):
+            ...
+            ValueError: the graph obtained after the addition of edge ((0, 2, None)) is not matching covered
+            sage: G.edges(sort=False) # No alteration to the existing graph
+            [(0, 1, None), (0, 3, None), (1, 2, None), (2, 3, None)]
+
+        Adding an edge with exactly one incident vertex that is nonexistent
+        throws a :exc:`ValueError` exception, as the resulting graph would
+        have an odd order::
+
+            sage: C = graphs.CycleGraph(4)
+            sage: G = MatchingCoveredGraph(C)
+            sage: G.add_edge(0, 4)
+            Traceback (most recent call last):
+            ...
+            ValueError: the graph obtained after the addition of edge ((0, 4, None)) is not matching covered
+            sage: G.edges(sort=False) # No alteration to the existing graph
+            [(0, 1, None), (0, 3, None), (1, 2, None), (2, 3, None)]
+
+        Adding an edge with both the incident vertices that is nonexistent
+        throws a :exc:`ValueError` exception, as the resulting graph would
+        have been disconnected::
+
+            sage: C = graphs.CycleGraph(4)
+            sage: G = MatchingCoveredGraph(C)
+            sage: G.add_edge(4, 5)
+            Traceback (most recent call last):
+            ...
+            ValueError: the graph obtained after the addition of edge ((4, 5, None)) is not matching covered
+            sage: G.edges(sort=False) # No alteration to the existing graph
+            [(0, 1, None), (0, 3, None), (1, 2, None), (2, 3, None)]
+        """
+        if label is None:
+            if v is None:
+                try:
+                    u, v, label = u
+                except Exception:
+                    try:
+                        u, v = u
+                    except Exception:
+                        pass
+
+        else:
+            if v is None:
+                try:
+                    u, v = u
+                except Exception:
+                    pass
+
+        if u in self and v in self:
+            if u == v:
+                raise ValueError('loops are not allowed in '
+                                 'matching covered graphs')
+            G = Graph(self)
```

Comment:
do we really need to create a new graph and check if it is matching covered ? can't we find a lighter incremental method ? 
At least, you can make add a TODO block at the top of the module to list what should be done to improve the situation.

Reply:
While adding a new edge to the graph, we need to check whether it lies in a perfect matching or not and that is in  the order of `|V| + |E|`. We are creating a new instance of the existing graph and after addition of the new edge, checking whether this new edge lies in a perfect matching or not using the current matching (this shall take linear time as we are using M alternating tree search method). So, even if we do not create a new instance of the graph, the overall complexity, in the worst case, is in linear.

Reply:
ok